samza-core/src/main/java/org/apache/samza/operators/impl/WatermarkStates.java

```diff
@@ -49,34 +50,59 @@ class WatermarkStates {
   private final static class WatermarkState {
     private final int expectedTotal;
     private final Map<String, Long> timestamps = new HashMap<>();
+    private final Map<String, Long> lastUpdateTime = new HashMap<>();
+    private final long watermarkIdleTime;
+    private final long createTime;
+    private final LongSupplier systemTimeFunc;
     private volatile long watermarkTime = WATERMARK_NOT_EXIST;
 
-    WatermarkState(int expectedTotal) {
+    WatermarkState(int expectedTotal, long watermarkIdleTime, LongSupplier systemTimeFunc) {
       this.expectedTotal = expectedTotal;
+      this.watermarkIdleTime = watermarkIdleTime;
+      this.systemTimeFunc = systemTimeFunc;
+      this.createTime = systemTimeFunc.getAsLong();
     }
 
     synchronized void update(long timestamp, String taskName) {
+      long currentTime = systemTimeFunc.getAsLong();
       if (taskName != null) {
         Long ts = timestamps.get(taskName);
         if (ts != null && ts > timestamp) {
           LOG.warn(String.format("Incoming watermark %s is smaller than existing watermark %s for upstream task %s",
               timestamp, ts, taskName));
         } else {
           timestamps.put(taskName, timestamp);
+          lastUpdateTime.put(taskName, currentTime);
         }
       }
 
       if (taskName == null) {
         // we get watermark either from the source or from the aggregator task
         watermarkTime = Math.max(watermarkTime, timestamp);
-      } else if (timestamps.size() == expectedTotal) {
-        // For any intermediate streams, the expectedTotal is the upstream task count.
-        // Check whether we got all the watermarks, and set the watermark to be the min.
-        Optional<Long> min = timestamps.values().stream().min(Long::compare);
-        watermarkTime = min.orElse(timestamp);
+      } else if (canUpdateWatermark(currentTime)) {
+        Optional<Long> min;
+        if (watermarkIdleTime <= 0) {
+          // All upstream tasks are required in the computation
+          min = timestamps.values().stream().min(Long::compare);
+        } else {
+          // Exclude the tasks that have been idle in watermark emission.
+          min = timestamps.entrySet().stream()
+                  .filter(t -> currentTime - lastUpdateTime.get(t.getKey()) < watermarkIdleTime)
```

Comment:
nit: two minor perf optimizations:
1. This will actually go through all the partitions multiple times, and will be slower than a single for-loop. (As SIMD probably won't help much here any ways).
2. `currentTime` - `watermarkIdleTime` is a constant, so it can be computed once and compared with the updateTime of each key.

Reply:
Makes sense. Changed to use a single loop.